crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+                )),
+            }
+        }
+    }
+
+    impl Display for Precision {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Precision {
+        fn default() -> Self {
+            Precision(MAX_DECIMAL_PRECISION)
+        }
+    }
+
+    // In Postgres, the scale can be between [-1000, 1000]
+    #[derive(
+        Debug,
+        Clone,
+        Copy,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        SizeOf,
+        rkyv::Archive,
+        rkyv::Serialize,
+        rkyv::Deserialize,
+    )]
+    #[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct Scale(u32);
+
+    impl Scale {
+        pub fn inner(&self) -> u32 {
+            self.0
+        }
+
+        pub fn min() -> Self {
+            Self(MIN_DECIMAL_SCALE)
+        }
+
+        pub const fn max() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+
+        pub fn apply(&self, val: &mut Decimal) {
+            val.rescale(self.0)
+        }
+    }
+
+    impl TryFrom<u32> for Scale {
+        type Error = String;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                x if (MIN_DECIMAL_SCALE..=MAX_DECIMAL_SCALE).contains(&x) => Ok(Scale(x)),
+                _ => Err(format!(
+                "Numeric scale {value} must be between {MIN_DECIMAL_SCALE} and {MAX_DECIMAL_SCALE}"
+            )),
+            }
+        }
+    }
+
+    impl Display for Scale {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Scale {
+        fn default() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+    }
+}
+
+use inner_types::*;
+
+/// A wrapper type around [`rust_decimal::Decimal`].
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, SizeOf)]
+#[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "with-serde", serde(transparent))]
+#[derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
+pub struct SQLDecimal {
+    inner: Decimal,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_precision: Precision,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_scale: Scale,
+}
+
+impl FromStr for SQLDecimal {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let inner = Decimal::from_str(s).map_err(|e| e.to_string())?;
+
+        Ok(Self {
+            inner,
+            max_precision: Precision::max(),
+            max_scale: Scale::try_from(inner.scale())?,
+        })
+    }
+}
+
+impl TryFrom<Decimal> for SQLDecimal {
+    type Error = String;
+
+    fn try_from(mut inner: Decimal) -> Result<Self, Self::Error> {
```

Comment:
When is this function used?

Reply:
This was used to create `SQLDecimal` internally from `Decimal` but will be removed in the next commit.